File: server/routes/coin.js
const { Router } = require("express");
const axios = require("axios");
const router = Router();
const { ETH } = require("../model/eth");
const { SOL } = require("../model/sol");

const coins = {
  eth: {
    name: "ethereum",
    symbol: "ETH",
    schema: ETH,
  },
  sol: {
    name: "solana",
    symbol: "SOL",
    schema: SOL,
  },
};

// Get Coin
router.get("/:coin", async (req, res) => {
  // Array of all the Coin currencies
  const currencyArr = [
    ["USD", "$"],
    ["EUR", "€"],
    ["BTC", "₿"],
    ["ILS", "₪"],
  ];

  // Coin obj
  let objCoin = {};

  // Coin Api url
  const url = `https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest?symbol=${req.params.coin}&convert=`;

  // My API key for the api
  const headers = {
    "X-CMC_PRO_API_KEY": "92644239-8cbd-47fa-b5d7-7b0ab1d60a9a",
  };

  // Loop that get all the Coin data for each currency and set them into objCoin object
  for (let i = 0; i < currencyArr.length; i++) {
    try {
      const { data } = await axios.get(url + currencyArr[i][0], { headers });

      // The currency name: USD/EUR/BTC
      const currency = currencyArr[i][0];

      // The currency symbol
      const symbol = currencyArr[i][1];

      const {
        price,
        market_cap,
        volume_24h,
        volume_change_24h,
        percent_change_1h,
        percent_change_24h,
        percent_change_7d,
        percent_change_30d,
        percent_change_60d,
        percent_change_90d,
        last_updated,
      } = data.data[coins[req.params.coin].symbol].quote[currency];
      if (!objCoin.last_updated) {
        objCoin.last_updated = new Date(last_updated).toLocaleString();
      }

      objCoin[currency] = {
        name: currency,
        symbol,
        price,
        market_cap,
        volume_24h,
        volume_change_24h,
        percent_change_1h,
        percent_change_24h,
        percent_change_7d,
        percent_change_30d,
        percent_change_60d,
        percent_change_90d,
      };
    } catch (error) {
      console.log(error.message);
      return res.status(400).send(error.message);
    }
  }

  res.status(200).send(objCoin);
  try {
    let coinToSave = new coins[req.params.coin].schema(objCoin);
    coinToSave = await coinToSave.save();
  } catch (error) {
    res.status(400).send(error.message);
  }
});

// Delete all
router.delete("/:coin", async (req, res) => {
  await coins[req.params.coin].schema.deleteMany();
  res.send("All deleted");
});

module.exports = router;

File: pyServer/routes/coin.py
from datetime import datetime

from fastapi import APIRouter, HTTPException
from pyServer.db import db
import httpx
from pyServer.models.eth import ETH
from pyServer.models.sol import SOL

from pyServer.models.currency import CurrencyData

router = APIRouter()

# Coin definitions
COINS = {
    "eth": {"symbol": "ETH", "collection": db.get_collection("eths"), "model": ETH},
    "sol": {"symbol": "SOL", "collection": db.get_collection("sols"), "model": SOL},
}

# Supported currencies
CURRENCIES = [
    ("USD", "$"),
    ("EUR", "€"),
    ("BTC", "₿"),
    ("ILS", "₪"),
]

CMC_API_KEY = "92644239-8cbd-47fa-b5d7-7b0ab1d60a9a"
CMC_URL = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest"


@router.get("/{coin}")
async def get_coin(coin: str):
    if coin not in COINS:
        raise HTTPException(status_code=404, detail="Coin not supported")

    coin_data = {}
    headers = {"X-CMC_PRO_API_KEY": CMC_API_KEY}

    async with httpx.AsyncClient() as client:
        for currency, symbol in CURRENCIES:
            try:
                r = await client.get(
                    CMC_URL,
                    params={"symbol": COINS[coin]["symbol"], "convert": currency},
                    headers=headers
                )
                data = r.json()["data"][COINS[coin]["symbol"]]["quote"][currency]

                coin_data[currency] = CurrencyData(
                    price=data["price"],
                    volume_24h=data["volume_24h"],
                    market_cap=data["market_cap"],
                    volume_change_24h=data["volume_change_24h"],
                    percent_change_1h=data["percent_change_1h"],
                    percent_change_24h=data["percent_change_24h"],
                    percent_change_7d=data["percent_change_7d"],
                    percent_change_30d=data["percent_change_30d"],
                    percent_change_60d=data["percent_change_60d"],
                    percent_change_90d=data["percent_change_90d"],
                    symbol=symbol,
                    name=currency
                )
            except Exception as e:
                raise HTTPException(status_code=400, detail=str(e))

    coin_model = COINS[coin]["model"](
        USD=coin_data.get("USD"),
        EUR=coin_data.get("EUR"),
        BTC=coin_data.get("BTC"),
        ILS=coin_data.get("ILS"),
        last_updated=datetime.now().isoformat()
    )

    collection = COINS[coin]["collection"]
    await collection.insert_one(coin_model.model_dump())

    return coin_model


@router.delete("/{coin}")
async def delete_coin(coin: str):
    if coin not in COINS:
        raise HTTPException(status_code=404, detail="Coin not supported")
    collection = COINS[coin]["collection"]
    await collection.delete_many({})
    return {"message": f"All {coin.upper()} documents deleted"}
